src/Common/src/TypeSystem/Common/TypeDesc.cs

```diff
@@ -465,5 +481,25 @@ public virtual MethodDesc GetFinalizer()
         {
             return null;
         }
+
+        /// <summary>
+        /// Gets a value indicating whether this type has generic variance (the definition of the type
+        /// has a generic parameter that is co- or contravariant).
+        /// </summary>
+        public bool HasVariance
+        {
+            get
+            {
+                return (GetTypeFlags(TypeFlags.HasGenericVariance | TypeFlags.HasGenericVarianceComputed) & TypeFlags.HasGenericVariance) != 0;
+            }
+        }
+
+        /// <summary>
+        /// Gets the kind of this type.
+        /// </summary>
+        public abstract TypeKind Variety
```

Comment:
The switch over `Category` becomes ugly for the `DefType` case because now it will be a `default` case with an `if`.

Also the fact that TypeFlags don't distinguish between SzArray and Array is mildly annoying.